### PR TITLE
Remove FilterByNames method to avoid overriding a pointer value

### PR DESCRIPTION
**Pull request by knanao**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 1]
```diff
@@ -97,30 +94,23 @@ type rbac struct {
...
+			us = append(us, v)
 		}
 	}
-	r.Roles = roles
-	return r
+	for _, v := range us {
+		if v.HasPermission(typ, action) {
+			return true
+		}
+	}
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:112]: how about this?
```
	for _, u := range userRoles {
		v, ok := rs[u]
                 if !ok {
			continue
		}
                 if v.HasPermission(typ, action) {
			return true
		}
	}
```

Comment (knanao) [reply pkg/app/server/service/webservice/service.pb.auth.go]: Great, Let me apply this.

File: pkg/model/project.go [hunk 2]
```diff
@@ -547,3 +547,19 @@ func (p *ProjectRBACPolicy) HasPermission(typ ProjectRBACResource_ResourceType,
 	}
 	return false
 }
+
+func VerifyRBACPermission(typ ProjectRBACResource_ResourceType, action ProjectRBACPolicy_Action, userRoles []string, roles []*ProjectRBACRole) bool {
```
Comment (nghialv) [inline pkg/model/project.go:551]: Now I understand why did you add a `Fillter...` function. 😄 
To avoid a complex parameter list of this function.

Comment (knanao) [reply pkg/model/project.go:551]: Actually, it was. lol
I intended to refactor it but ended up with code that produces bugs.

Comment (nghialv) [reply pkg/model/project.go:551]: Just an idea, how about doing the filter inside `getRBACRoles` function and returning only the matched list?
The signature will be ` getRBACRoles(ctx context.Context, projectID string, roles []string) ([]*model.ProjectRBACRole, error)` and then we can use the returned list directly.
It is more nature. 🤔

Comment (khanhtc1202) [reply pkg/model/project.go:551]: to reduce the complex parameter list, there is another way which is to make this verify function as closure in the `Authorize` method, then we can remove `userRoles []string & roles []*ProjectRBACRole` from the closure function signature, but it's up to you. TBH, I don't think making this VerifyRBACPermission function as part of the model package has much meaning 😄

Comment (knanao) [reply pkg/model/project.go:551]: That's a good idea. Please apply these changes.

Comment (knanao) [reply pkg/model/project.go:551]: That's a good idea. Please apply these changes.

Comment (knanao) [reply pkg/model/project.go:551]: > I don't think making this VerifyRBACPermission function as part of the model package has much meaning 
I think it's better to write the verifying logic in the model pkg because `xxx.auth.go` file is automatically generated and it's not easy to change. Hence I wanna write some logic into another stuff as much as possible. wdyt?

Comment (khanhtc1202) [reply pkg/model/project.go:551]: > xxx.auth.go file is automatically generated and it's not easy to change.

I get your point and feel the same too. But I prefer encapsulation-based codebase design, which means if the code does not actually be used in multiple places, should not move it to a common package (the model in this case). The auth codebase is autogenerated means even if we want to update it later, it should only be affected by itself template code as much as possible (reduce the risk if we change code in the common package which affect generated code).

Comment (knanao) [reply pkg/model/project.go:551]: Oh, I see and that makes sense.
Alright, let's write `VerifyRBACPermission` in the same file.

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 3]
```diff
@@ -60,19 +59,16 @@ func NewRBACAuthorizer(
 	}
 }
 
-func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, error) {
+func (a *authorizer) getRBACRoles(ctx context.Context, projectID string, roles []string) ([]*model.ProjectRBACRole, error) {
 	if _, ok := a.projectsInConfig[projectID]; ok {
 		p := &model.Project{Id: projectID}
 		p.SetBuiltinRBACRoles()
-		return &rbac{p.RbacRoles}, nil
+		return p.RbacRoles, nil
```
Comment (nghialv) [inline pkg/app/server/service/webservice/service.pb.auth.go:66]: Do we need to filter this before returning?

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 4]
```diff
@@ -60,19 +59,16 @@ func NewRBACAuthorizer(
...
 	}
 
-	a.logger.Info("call getting a rbac cache")
-	r, err := a.rbacCache.Get(projectID)
-	if err == nil {
-		return r.(*rbac), nil
+	if v, err := a.rbacCache.Get(projectID); err == nil {
+		return v.([]*model.ProjectRBACRole), nil
```
Comment (nghialv) [inline pkg/app/server/service/webservice/service.pb.auth.go:70]: Do we need to filter this before returning?

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 5]
```diff
@@ -83,156 +79,141 @@ func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, erro
...
 	}
-	r.Roles = roles
-	return r
+	return false
 }
 
 // Authorize checks whether a role is enough for given gRPC method or not.
 func (a *authorizer) Authorize(ctx context.Context, method string, r model.Role) bool {
-	rbac, err := a.getRBAC(ctx, r.ProjectId)
+	roles, err := a.getRBACRoles(ctx, r.ProjectId, r.ProjectRbacRoles)
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:113]: this `getRBACRoles` is a bit odd to me, the return is not RBACRoles but actually `r.ProjectRbacRoles` or ProjectRBACRoles, so how about renaming it to something like `getProjectRbacRoles`?

Comment (knanao) [reply pkg/app/server/service/webservice/service.pb.auth.go:113]: @nghialv @khanhtc1202 
how about like this?
```
func (a *authorizer) getProjectRBACRoles(ctx context.Context, projectID string, roles []string) ([]*model.ProjectRBACRole, error) {
	all, err := a.getAllProjectRBACRoles(ctx, projectID)
	if err != nil {
		return nil, err
	}
	rs := make(map[string]*model.ProjectRBACRole, len(all))
	for _, v := range all {
		rs[v.Name] = v
	}
	ret := make([]*model.ProjectRBACRole, 0, len(roles))
	for _, r := range roles {
		if v, ok := rs[r]; ok {
			ret = append(ret, v)
		}
	}
	return ret, nil
}

func (a *authorizer) getAllProjectRBACRoles(ctx context.Context, projectID string) ([]*model.ProjectRBACRole, error) {
	if _, ok := a.projectsInConfig[projectID]; ok {
		p := &model.Project{Id: projectID}
		p.SetBuiltinRBACRoles()
		return p.RbacRoles, nil
	}

	if v, err := a.rbacCache.Get(projectID); err == nil {
		return v.([]*model.ProjectRBACRole), nil
	}

	p, err := a.projectStore.Get(ctx, projectID)
	if err != nil {
		a.logger.Error("failed to get project",
			zap.String("project", projectID),
			zap.Error(err),
		)
		return nil, err
	}

	if err = a.rbacCache.Put(projectID, p.RbacRoles); err != nil {
		a.logger.Warn("unable to store the rbac in memory cache",
			zap.String("project", projectID),
			zap.Error(err),
		)
	}
	return p.RbacRoles, nil
}
```

Comment (khanhtc1202) [reply pkg/app/server/service/webservice/service.pb.auth.go]: LGTM 👍

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 6]
```diff
@@ -83,156 +97,137 @@ func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, erro
...
-type rbac struct {
-	Roles []*model.ProjectRBACRole
+	return p.RbacRoles, nil
 }
 
-func (r *rbac) HasPermission(typ model.ProjectRBACResource_ResourceType, action model.ProjectRBACPolicy_Action) bool {
-	for _, v := range r.Roles {
-		if v.HasPermission(typ, action) {
-			log.Printf("this rbac has a permission: %v", r.Roles)
+func verifyRBACPermission(typ model.ProjectRBACResource_ResourceType, action model.ProjectRBACPolicy_Action, roles []*model.ProjectRBACRole) bool {
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:109]: nits, this can be Authorize's closure to reduce passing `roles` every call.

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 7]
```diff
@@ -83,156 +97,137 @@ func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, erro
...
 
-	a.logger.Info("success to get the rbac", zap.Any("rbac", rbac))
-	rbac.FilterByNames(r.ProjectRbacRoles)
-	a.logger.Info("sucess to filter the rbac", zap.Any("filtered-rbac", rbac))
+	if len(roles) == 0 {
+		a.logger.Error("the user does not have existing RBAC roles",
+			zap.String("project", r.ProjectId),
+		)
+		return false
+	}
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:134]: should be handled in `verifyRBACPermission` function?

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 8]
```diff
@@ -83,156 +97,137 @@ func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, erro
...
 		)
 		return false
 	}
 
-	a.logger.Info("success to get the rbac", zap.Any("rbac", rbac))
-	rbac.FilterByNames(r.ProjectRbacRoles)
-	a.logger.Info("sucess to filter the rbac", zap.Any("filtered-rbac", rbac))
+	if len(roles) == 0 {
+		a.logger.Error("the user does not have existing RBAC roles",
+			zap.String("project", r.ProjectId),
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:122]:
```suggestion
			zap.String("project", r.ProjectId),
			zap.Strings("userRoles", r.ProjectRbacRoles),
```
nits

File: pkg/app/server/service/webservice/service.pb.auth.go [hunk 9]
```diff
@@ -83,156 +97,137 @@ func (a *authorizer) getRBAC(ctx context.Context, projectID string) (*rbac, erro
...
+		)
+		return false
+	}
+
+	verifyRBACPermission := func(typ model.ProjectRBACResource_ResourceType, action model.ProjectRBACPolicy_Action) bool {
```
Comment (khanhtc1202) [inline pkg/app/server/service/webservice/service.pb.auth.go:127]:
```suggestion
	verify := func(typ model.ProjectRBACResource_ResourceType, action model.ProjectRBACPolicy_Action) bool {
```
nits

**Review by khanhtc1202 — APPROVED** [review]
Nice work 🙌

**Review by nghialv — APPROVED** [review]
Looks nice.
